### PR TITLE
Refactor catalog utils (again)

### LINKED ISSUE #152: Better catalog-related documentation
### Addressing a Problem?

The catalog notebook could be improved, especially the section on `parse_directory`, to better assist people who want to create a catalog for their own datasets.

### Potential Solution

_No response_

### Additional context

_No response_

### Contribution

- [ ] I would be willing/able to open a Pull Request to contribute this feature.

### PR DESCRIPTION
**Pull request by aulemahal**
### Pull Request Checklist:
- [x] This PR addresses an already opened issue (for bug fixes / features)
    - This PR fixes #152.
- [x] (If applicable) Documentation has been added / updated (for bug fixes / features)
- [x] This PR does not seem to break the templates.
- [x] HISTORY.rst has been updated (with summary of main changes)
  - [ ] Link to issue (:issue:`number`) and pull request (:pull:`number`) has been added

### What kind of change does this PR introduce?

* Move catalog creation (`parse_directory`) to a new `catutils.py` module
* Rewrite file parsing functions:
    - Use `os.walk` instead of subprocess calling  `find`. The function should now be platform independent and not linux-specific.
    - Use `parse` instead of `intake.source.utils.reverse_format`. It allows some interesting features, like format specifiers.
    - Rethink parallelism.
    - More flexibility to `cvs`
* Copy and adapt path building stuff from miranda. See `build_path`. Schema is `data/file_schema.yml`.
* `date_parser` has moved to `utils.py`.

### Does this PR introduce a breaking change?
Yes.

- Functions have moved.
- Patterns feeded to `parse_directory` are now specified differently. See doc and example below.
- No more `globpattern`. The extension is parsed from the patterns. Calls can now mix different extensions. A new `dirglob` arg takes care of the "folder filtering" feature.
- `parse_directory` does not use `dask` anymore. I believe the code is now easier to understand and speed is not affected too much. Parsing a local filesystem is not supposed to be faster in parallel, but we are often using nfs-mounted disk where this is less true. Testing on doris, parsing the full `/datasets/simulation/raw` folder takes the same time between the new and the old versions. See note below on "code complexity".

### Other information:
Example for the new patterns:

```
OLD: {processing_level}/{mip_era}/{activity}/{domain}/{institution}/{source}/{experiment}/{member}/{frequency}/{_variable}/{?var}_{?freq}_{?src}_{?exp}_{?memb}_{*}_{date_start}-{date_end}.nc"

NEW : "{processing_level}/{mip_era}/{activity}/{domain}/{institution}/{source}/{experiment}/{member}/{frequency}/{variable:_}/{*}_{DATES}.nc"
```
The "variable" field accepts underscores, so it uses the "_" format specifier.  No more need to specify each part of the filename if only the last one is needed. Here the "DATES" special field can catch single dates or bounds. 

### TODO:
- ~Update the notebooks~

### Code complexity
Lol. I thought this PR would simplify the `parse_directory` system.

The problem is the MRCC5. Or at least, it is the enormous size of this database and its scattering on slow-to-read disks. It is the only reason for all these complexities:
- Queues to multithread the file parsing
- Access checks because there are non-readable files in the database.
- Multiprocessing pool for parallelize the disks parsing
- "read_file" by groups to minimize the number of file reads.
- Parsing from file with fast-paths for netCDF and Zarr (instead of xarray).
- Nested CVS (although this new feature could be useful in other databases)

I wanted to make clear that the complexity of this code is NOT only because I like to optimize things. Last time I ran the MRCC5 catalog creation (with this code), it took 7 hours. Just imagine without the optimizations. (And this while missing a full disk).

### CODE REVIEW

**Comment by RondeauG** [comment]
If you could use this PR to also address #152, that would be great! I wrote the issue, but it originally came from @mccrayc, so you can check with him for the details.

File: xscen/data/file_schema.yml [hunk 1]
```diff
@@ -0,0 +1,192 @@
...
+#
+#     - domain denotes the zone over which the scenario was computed, not the original source's coverage
+#     - processing_level has more possible values, denoting slight variant of the data instead of broad categories.
+#     As such, both fields are here much lower in the hierarchy.
+derived-sims:
+  with:
+  - option: type
+    value: simulation
+  folders:
+    - type
```
Comment (juliettelavoie) [inline xscen/data/file_schema.yml:120]: Where will does go ? If we just put them in /datasets like that, we will have:
|_ simulation
            |_ raw
            |_ biasadjusted
            |_ ESPO-G6-R2_v1.0 #containing indicators and other derived...

If it stays in /datasets, we could add "derived" after type? 
Or we just know that we have to put the root somewhere else ?

**Review by juliettelavoie — COMMENTED** [review]
This is great!

1. I would update the init to be able to do `xs.build_path`.
2. It might be good to add `build_path` at the end of the the Getting Started notebook.
3. There is some conflict with format and version.

```
ds = xr.open_zarr('PATH_FINAL_ESPO')
ds = ds[['tasmax']]
ds.attrs['cat:variable']='tasmax'
ds.attrs['cat:processing_level']='biasadjusted'
xs.catutils.build_path(ds) #--> PosixPath('simulation/biasadjusted/ESPO-G6_1.1.0/CMIP6/ScenarioMIP/NAM-rdrs/BCC/BCC-CSM2-MR/ssp245/r1i1p1f1/day/tasmax/tasmax_day_ESPO-G6_1.1.0_CMIP6_ScenarioMIP_NAM-rdrs_BCC_BCC-CSM2-MR_ssp245_r1i1p1f1_1950-2100')
xs.catutils.build_path(ds,format='zarr') #-->PosixPath('simulation/ESPO-G6_1.1.0/CMIP6/ScenarioMIP/BCC/BCC-CSM2-MR/ssp245/r1i1p1f1/NAM-rdrs/final/D/tasmax/tasmax_D_ESPO-G6_1.1.zarr')
```
4. In an issue, we discussed moving some of my ESPO utils fonction in xscen. I think the `save_move_update` doesn't have a place anymore with the new high quotas and io issues we are having. But I think a `save_and_update` with possibility of using `build_path` would be nice. It would lighten workflows. I could push this on this PR or a new one after.

File: xscen/catutils.py [hunk 2]
```diff
@@ -0,0 +1,1071 @@
...
+        List of column names on which to base the dataset definition. Empty columns will be skipped.
+        If None (default), it uses :py:data:`ID_COLUMNS`.
+    read_from_file : boolean or set of strings or tuple of 2 sets of strings.
+        If True, if some fields were not parsed from their path, files are opened and
+        missing fields are parsed from their metadata, if found.
+        If a sequence of column names, only those fields are parsed from the file, if missing.
+        If False (default), files are never opened.
+        If a tuple of 2 lists of strings, only the first file of groups defined by the
+        first list of columns is read and the second list of columns is parsed from the
+        file and applied to the whole group.
```
[the remark this answers is not in the document]
Comment (aulemahal) [reply xscen/catutils.py:447]: This is correct.

File: xscen/catutils.py [hunk 3]
```diff
@@ -0,0 +1,1077 @@
...
+        and start.day == 1
+        and (
+            freq >= pd.Timedelta(CV.xrfreq_to_timedelta("YS"))
+            or (end.month == 12 and end.day > 29)
+        )
+    ):
+        if start.year == end.year:
+            return f"{start:%Y}"
+        return f"{start:%Y}-{end:%Y}"
+    # Full months : Starts on the 1st and is either montly or ends on the last day
```
Comment (RondeauG) [inline xscen/catutils.py:890]:
```suggestion
    # Full months : Starts on the 1st and is either monthly or ends on the last day
```

File: xscen/io.py [hunk 4]
```diff
@@ -20,17 +20,47 @@
...
+    copy: bool
+        If True (default), the data is copied. If False, it is moved.
+        If path_out is a dict, the data is copied then removed, otherwise it depends on the behaviour of :py:func:`shutil.move`.
+    overwrite: bool
+        If the destination already exists, an error is raised if overwrite is False (default).
+    erase_on_error: bool
+        If True (default), copy is True and the copy fails, the destination is erased before reraising the error.
+        Has no effect if copy is False.
+    """
+    pass
```
Comment (RondeauG) [inline xscen/io.py:61]: We should have a `NotImplementedError`

**Review by RondeauG — APPROVED** [review]
Looks good! However, I'd ideally wait until @mccrayc has had a look at it before we merge.

File: xscen/catutils.py [hunk 5]
```diff
@@ -0,0 +1,1077 @@
...
+        first list of columns is read and the second list of columns is parsed from the
+        file and applied to the whole group. For example, `(["source"],["institution", "activity"])` will find a
+        group with all the files that have the same source, open only one of the files to read the institution
+        and activity, and write this information in the catalog for all filles of the group.
+        It can also be a list of those tuples.
+    homogenous_info : dict, optional
+        Using the {column_name: description} format, information to apply to all files.
+        These are applied before the `cvs`.
+    cvs: str or PosixPath or dict, optional
+        Dictionary with mapping from parsed term to preffered terms (Controlled VocabularieS) for each column.
```
[the remark this answers is not in the document]
Comment (juliettelavoie) [reply xscen/catutils.py:455]: the goal of the captial S is to map to CVS

**Review by juliettelavoie — APPROVED** [review]

**Comment by mccrayc** [comment]
> 

Looks good to me! The new patterns seem much more intuitive and clean.

**Comment by aulemahal** [comment]
Last commit made a few changes. I tested the new code with ouranos_data_catalogs and it made a few bugs appear:

- `_parse_from_zarr` can now read the time coordinate. Before, xarray was needed, which was slowing down the process a lot.
- Removed `split_dataset`, this is to be implemented in another PR.
- Relaxed the "dates" regex to dates of any length.
- Added a test in `parse_directory` to remove entries where  `date_start` is a date, but `frequency` is "fx".

And I guess I now need to add tests!

**Comment by aulemahal** [comment]
+12% coverage :muscle:

**Pull request merged**